feat(mme): NGAP Unit Test CMake changes and addition of new cases

panyogesh opened the pull request:
Changes:
  1.  Cmake changes for ngap unit testing directory.
  2.  Renamed the existing files to show the coverage area.
  3.  Added new cases for handling new connection
 
Test:
         1. Unit Testing (test_oai)
             Test Logs :
            UnitTest  Logs
               6/18 Test  #6: test_ngap .........................   Passed    0.19 sec
   
      Call setup Logs
        MME Logs:
         000633 Wed Oct 06 02:20:41 2021 7F6B155E9700 DEBUG NGAP   tasks/ngap/ngap_client_servicer.:0044       
          Is Mock Enabled 0

      UERANSIM Logs:
          [2021-10-06 02:06:13.973] [ngap] [info] PDU session resource(s) setup for UE[1] count[1]
          [2021-10-06 02:06:13.997] [app] [info] Connection setup for PDU session[1] is successful, TUN 
          interface[uesimtun0, 192.168.128.254] is up.

Signed-off-by: Yogesh Pandey <yogesh@wavelabs.ai>

<!--
    Tag your PR title with the components that it touches along with
    the type of change
    E.g. "fix(orc8r): Fix reindexer race condition" or "feat(agw) ..."
-->

## Summary

<!-- Enumerate changes you made and why you made them -->

## Test Plan

<!--
    How did you test your change? How do you know it works?
    Add supporting screenshots, terminal pastes, etc. as necessary
-->

## Additional Information

- [ ] This change is backwards-breaking

<!--
    If this is a backwards-breaking change, document the upgrade instructions.
    All upgrade instructions for backwards-breaking changes will be aggregated
    in the next release's changelog so this is very important to fill out.
-->

panyogesh committed:
feat(mme): NGAP Unit Testing Setup Changes

Changes:
  1. Updated the Cmake fir test/ngap
  2. Renamed the test_ngap file to reflect the
     intent of functionality

Test:
  1. Unit Testing (test_oai)

Signed-off-by: Yogesh Pandey <yogesh@wavelabs.ai>

ulaskozat reviewed (changes requested):
This adds a third pattern in unit testing. Currently we use gmock and `#if !MME_UNIT_TEST` patterns. I think we should not have a runtime logic to check whether this is unit testing or not. cc @ardzoht @markjen

panyogesh commented:
> This adds a third pattern in unit testing. Currently we use gmock and `#if !MME_UNIT_TEST` patterns. I think we should not have a runtime logic to check whether this is unit testing or not. cc @ardzoht @markjen

Sure.. I will change it to compile time.

panyogesh committed:
feat(mme): Added the review comments

 1. Changed mock option to run time instead of compile time.
 2. Add one testcase to check the flow towards AMF

Signed-off-by: Yogesh Pandey <yogesh@wavelabs.ai>

ulaskozat commented on lte/gateway/c/core/oai/test/ngap/test_ngap_flows.cpp:
```diff
@@ -0,0 +1,109 @@
+      0x00, 0x22, 0x42, 0x65, 0x00, 0x00, 0x01, 0xe4, 0xf7, 0x04, 0x44,
+      0x00, 0x5a, 0x40, 0x01, 0x18, 0x00, 0x70, 0x40, 0x01, 0x00};
+
+  itti_init(
+      TASK_MAX, THREAD_MAX, MESSAGES_ID_MAX, tasks_info, messages_info, NULL,
+      NULL);
```
is this still needed if there are no actual itti messages being passed among the threads?

panyogesh replied:
itti_init() is needed as it fills the MessageDef struct. For instance in func _ngap_amf_itti_ngap_initial_ue_message()_ a new message is allocated (NGAP_INITIAL_UE_MESSAGE). Without itti_init indexing can't be done for _NGAP_INITIAL_UE_MESSAGE_ and hence will result in crash.

ulaskozat commented on lte/gateway/c/core/oai/tasks/ngap/ngap_amf.c:
```diff
@@ -377,5 +379,5 @@ status_code_e ngap_send_msg_to_task(
       ITTI_MSG_NAME(message));
 
-  return send_msg_to_task(task_zmq_ctx_p, destination_task_id, message);
+  return send_message_to_amf(task_zmq_ctx_p, destination_task_id, message);
```
why not instead move `ngap_send_msg_to_task` to c++ file and have the content of `send_message_to_amf` in there?

panyogesh replied:
Comments addressed.  Removed the client_enforcer layer along with send_message_to_amf() funciton.

ulaskozat commented on lte/gateway/c/core/oai/tasks/ngap/ngap_client_enforcer.cpp:
```diff
@@ -0,0 +1,25 @@
+    MessageDef* message) {
+  return (NGAPClientServicer::getInstance().send_message_to_amf(
+      task_zmq_ctx_p, destination_task_id, message));
+}
```
why can't this be merged with `ngap_client_servicer.cpp`?

panyogesh replied:
Comments addressed.  Removed the client_enforcer layer.

ulaskozat commented on lte/gateway/c/core/oai/tasks/ngap/include/ngap_client_enforcer.h:
```diff
@@ -0,0 +1,29 @@
+
+#ifdef __cplusplus
+}
+#endif
```
why can't this be merged with `ngap_client_servicer.h`?

panyogesh replied:
Comments addressed.  Removed the client_enforcer layer.

panyogesh committed:
feat(mme): Review comments for ngap UT (#9414)

Changes :
  1. Removed the client_enforcer files
  2. Used ngap_send_msg_to_task as the main task
     which can be over-ridden

Tests:
     1. Unit testing
     2. Basic call flows with simulators

Signed-off-by: Yogesh Pandey <yogesh@wavelabs.ai>

ulaskozat commented on lte/gateway/c/core/oai/test/ngap/test_ngap_flows.cpp:
```diff
@@ -0,0 +1,109 @@
+  EXPECT_EQ(state->gnbs.num_elements, 1);
+
+  Ngap_NGAP_PDU_t decoded_pdu = {};
+  unsigned short length =
```
legit github action comment

panyogesh replied:
Comments addressed.

ulaskozat reviewed (approved):
LGTM

ulaskozat commented:
@panyogesh there is one minor nit. please address it and then mark this as ready to merge.

panyogesh committed:
feat(mme): Addressing github review comment

Signed-off-by: Yogesh Pandey <yogesh@wavelabs.ai>

ulaskozat reviewed (approved):
LGTM
